### PR TITLE
[examples] Generate argparsers from type hints on dataclasses

### PR DESCRIPTION
**Pull request by julien-c**

### CODE REVIEW

**Review by thomwolf — COMMENTED** [review]
I really like the approach of using `dataclasses` and parsing arguments based on the types of the field. It's very clean.

But I find the `HFArgumentParser` class a bit confusing and I don't think I would be able to use it without checking the internals.

Here is a proposal for discussion: why don't we make `HFArgumentParser` directly inherit from python standard `ArgumentParser` instead of wrapping it? Everybody knows `ArgumentParser`. Then, the `pipeline` method (see my comments) could be an `HFArgumentParser.from_dataclass_list` constructor and the `add_arguments` could be an additional `add_argument_from_dataclass` method.

File: src/transformers/hf_argparser.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    parser: ArgumentParser
+
+    def __init__(self, parser: Optional[ArgumentParser] = None):
+        if parser is not None:
+            self.parser = parser
+        else:
+            self.parser = ArgumentParser()
+
+    def add_arguments(self, obj: DataClass) -> ArgumentParser:
```
[the remark this answers is not in the document]
Comment (thomwolf) [reply src/transformers/hf_argparser.py:24]: or `add_dataclass_arguments`?

File: examples/run_language_modeling.py [hunk 2]
```diff
@@ -532,75 +534,26 @@ def main():
...
+    training_args = TrainingArgs()
+    HfArgparser().add_arguments(training_args).parse_args(remaining_args, namespace=training_args)
+
+    # For now, let's merge the two sets of args into one,
+    # but soon, we'll keep distinct sets of args, with a cleaner separation.
+    args = argparse.Namespace(**vars(dataprocessing_args), **vars(training_args))
+    # Soon, we'll do this instead:
+    # dataprocessing_args = DataprocessingArgs()
+    # training_args = TrainingArgs()
+    # HfArgparser.pipeline([dataprocessing_args, training_args])
```
Comment (LysandreJik) [inline examples/run_language_modeling.py:556]: This would be very clean!

File: examples/run_language_modeling.py [hunk 3]
```diff
@@ -532,75 +534,26 @@ def main():
...
-        "See details at https://nvidia.github.io/apex/amp.html",
-    )
-    parser.add_argument("--local_rank", type=int, default=-1, help="For distributed training: local_rank")
-    parser.add_argument("--server_ip", type=str, default="", help="For distant debugging.")
-    parser.add_argument("--server_port", type=str, default="", help="For distant debugging.")
-    args = parser.parse_args()
+    dataprocessing_args, remaining_args = parser.parse_known_args()
+
+    training_args = TrainingArgs()
+    HfArgparser().add_arguments(training_args).parse_args(remaining_args, namespace=training_args)
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply examples/run_language_modeling.py:548]: I agree that I definitely wouldn't remember how to use this line.

**Review by LysandreJik — APPROVED** [review]
I really like how all arguments are centralized, and how this would simplify the code in examples! I agree with Thomas' comments, but overall it's LGTM for me.

File: tests/test_hf_argparser.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+class HfArgparserTest(unittest.TestCase):
+    def argparsersEqual(self, a: argparse.ArgumentParser, b: argparse.ArgumentParser) -> bool:
+        """
+        Small helper to check pseudo-equality of parsed arguments on `ArgumentParser` instances.
+        """
+        self.assertEqual(len(a._actions), len(b._actions))
+        for x, y in zip(a._actions, b._actions):
+            xx = {k: v for k, v in vars(x).items() if k != "container"}
+            yy = {k: v for k, v in vars(y).items() if k != "container"}
+            self.assertEqual(xx, yy)
```
Comment (sshleifer) [inline tests/test_hf_argparser.py:58]: `assertDictEqual` has nice tracebacks

File: src/transformers/hf_argparser.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sshleifer) [reply src/transformers/hf_argparser.py:24]: IIUC, this is sometimes called with an instance like `TrainingArgs()` and other times a type (like `BasicExample`), so the type annotation confused me.
Cool that `dataclasses.fields` works on both, though!

**Review by sshleifer — APPROVED** [review]
Very cool! +1 to the inheritance idea.

I wonder what happens in collision cases where an argparser already has an argument e.g. `batch_size`. It could be (a) correct type, (b) incorrect type. (c) correct type but a positional arg (e.g. user already added  `batch_size` now adds `--batch_size`.

File: examples/run_glue.py [hunk 5]
```diff
@@ -36,6 +36,8 @@
     AutoConfig,
     AutoModelForSequenceClassification,
     AutoTokenizer,
+    HfArgparser,
```
[the remark this answers is not in the document]
Comment (julien-c) [reply examples/run_glue.py]: (can always change it if needed)

File: examples/run_language_modeling.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (julien-c) [reply examples/run_language_modeling.py]: Yep, way simpler now

File: examples/run_language_modeling.py (hunk 2 above)
Comment (julien-c) [reply examples/run_language_modeling.py]: 🤗 🤗

**Comment by julien-c** [comment]
Phew! Ok, I went through multiple rewrites today and I think it is pretty good now.

**TL;DR:**

- I only made changes to the `run_glue.py` example script to keep the diff smaller.
- I have to pass `DataClassType`s (_types_, not instances) to `HfArgumentParser` because if they have required properties/arguments, we wouldn't be able to instantiate them before "filling" them
- The class is designed to play well with the native `argparse`. In particular, you can get back any not-known args and parse them using a different argparse.ArgumentParser, to make adoption easier in complex scripts.
- **read the unit tests for (a subset of) the supported arguments and how the properties translate into arguments.**

File: examples/run_glue.py [hunk 6]
```diff
@@ -383,137 +387,60 @@ def load_and_cache_examples(args, task, tokenizer, evaluate=False):
...
     )
-    parser.add_argument(
-        "--model_type",
-        default=None,
-        type=str,
-        required=True,
-        help="Model type selected in the list: " + ", ".join(MODEL_TYPES),
+    model_type: str = field(metadata={"help": "Model type selected in the list: " + ", ".join(MODEL_TYPES)})
+    model_name_or_path: str = field(
+        metadata={"help": "Path to pre-trained model or shortcut name selected in the list: " + ", ".join(ALL_MODELS)}
```
Comment (thomwolf) [inline examples/run_glue.py:401]: Duplicate `model_name_or_path` I think

File: examples/run_glue.py [hunk 7]
```diff
@@ -383,137 +387,60 @@ def load_and_cache_examples(args, task, tokenizer, evaluate=False):
...
-        "--output_dir",
-        default=None,
-        type=str,
-        required=True,
-        help="The output directory where the model predictions and checkpoints will be written.",
+    cache_dir: Optional[str] = field(
+        default=None, metadata={"help": "Where do you want to store the pre-trained models downloaded from s3"}
     )
+    do_lower_case: bool = field(default=False, metadata={"help": "Set this flag if you are using an uncased model."})
+    # ^ I want to remove this one because this should be handled by the tokenizer now.
```
Comment (thomwolf) [inline examples/run_glue.py:413]: Yes! You can remove it I think (cc @LysandreJik)

File: examples/run_glue.py [hunk 8]
```diff
@@ -383,137 +387,60 @@ def load_and_cache_examples(args, task, tokenizer, evaluate=False):
...
-        help="Whether to use 16-bit (mixed) precision (through NVIDIA apex) instead of 32-bit",
-    )
-    parser.add_argument(
-        "--fp16_opt_level",
-        type=str,
-        default="O1",
-        help="For fp16: Apex AMP optimization level selected in ['O0', 'O1', 'O2', and 'O3']."
-        "See details at https://nvidia.github.io/apex/amp.html",
+def main():
+    model_args, dataprocessing_args, training_args = HfArgumentParser.parse_into_dataclasses(
```
Comment (thomwolf) [inline examples/run_glue.py:437]: `parse_from_dataclasses`?

File: src/transformers/hf_argparser.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            else:
+                kwargs["type"] = field.type
+                if field.default is not dataclasses.MISSING:
+                    kwargs["default"] = field.default
+                else:
+                    kwargs["required"] = True
+            self.add_argument(field_name, **kwargs)
+
+    @classmethod
+    def parse_into_dataclasses(
```
Comment (thomwolf) [inline src/transformers/hf_argparser.py:69]: `parse_args_into_dataclasses`?

File: src/transformers/hf_argparser.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+DataClassType = NewType("DataClassType", Any)
+
+
+class HfArgumentParser(ArgumentParser):
+    """
+    This subclass of `argparse.ArgumentParser` uses type hints on dataclasses
+    to generate arguments.
+    """
+
+    def __init__(self, obj: Union[None, DataClass, DataClassType] = None):
```
Comment (thomwolf) [inline src/transformers/hf_argparser.py:17]: I would let it accept a list of DataClass as well store these in a `self.dataclasse_types` list (in the `add_dataclass_arguments ` method) to be able to generate the dataclasses when parsing like in the class method `parse_into_dataclasses`.

File: src/transformers/hf_argparser.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        Args:
+            obj:
+                (Optional) Can be either a dataclass instance, or a dataclass type.
+                Both will work the exact same way.
+        """
+        super().__init__()
+        if obj is not None:
+            self.add_dataclass_arguments(obj)
+
+    def add_dataclass_arguments(self, obj: Union[DataClass, DataClassType]):
```
Comment (thomwolf) [inline src/transformers/hf_argparser.py:28]: Maybe accept a list of Dataclasses as well?

Name could be also `add_arguments_from_dataclasses`

Comment (thomwolf) [reply src/transformers/hf_argparser.py:28]: Store the input Type in a `self.dataclasse_types` to generate dataclasses when parsing?

File: src/transformers/hf_argparser.py (hunk 9 above)
Comment (thomwolf) [reply src/transformers/hf_argparser.py:69]: I see the `classmethod` is in the "pipeline" spirit but if you want to have a standard argparse way to use this, this could be an instance method relying on a `self.dataclasse_types` attribute storing the dataclasses.

**Review by thomwolf — APPROVED** [review]
A few proposals to make it more similar to ArgParse, otherwise looks good!

This should be added in python 3.9 IMO :-)

File: src/transformers/hf_argparser.py (hunk 9 above)
Comment (thomwolf) [reply src/transformers/hf_argparser.py:69]: cf my example on `run_glue.py`

File: examples/run_glue.py (hunk 8 above)
Comment (julien-c) [reply examples/run_glue.py:437]: Yes, I almost proposed that variant yesterday :)

I'd even do:
```python
parser = HfArgumentParser([ModelArguments, DataProcessingArguments, TrainingArguments])
model_args, dataprocessing_args, training_args = parser.parse_args_into_dataclasses()
```

File: examples/run_glue.py (hunk 7 above)
Comment (LysandreJik) [reply examples/run_glue.py:413]: Indeed!

Comment (LysandreJik) [reply examples/run_glue.py:413]: Should update the documentation as well to remove those

File: src/transformers/hf_argparser.py [hunk 12]
```diff
@@ -12,27 +12,32 @@ class HfArgumentParser(ArgumentParser):
...
-        for field in dataclasses.fields(obj):
+        super().__init__(**kwargs)
+        if dataclasses.is_dataclass(dataclass_types):
+            dataclass_types = [dataclass_types]
+        self.dataclass_types = dataclass_types
+        for dtype in self.dataclass_types:
+            self._add_dataclass_arguments(dtype)
+
+    def _add_dataclass_arguments(self, dtype: DataClassType):
+        for field in dataclasses.fields(dtype):
```
Comment (julien-c) [inline src/transformers/hf_argparser.py:40]: Ok, final set of improvements before merging! 
- Everything is done at init time.
- You can add_argument using the standard method (not dataclasses backed) on the parser and you'll get the resulting object at the end of the tuple when calling `parse_args_into_dataclasses()`
